File: js/endless-binary/power-of-two-1.js
// taidalab Version 1.5.2
// https://github.com/taidalog/taidalab
// Copyright (c) 2022 taidalog
// This software is licensed under the MIT License.
// https://github.com/taidalog/taidalab/blob/main/LICENSE
function checkAnswerPot1 (answer, hint_format, last_answers) {
    // Getting the user input.
    const numberInput = document.getElementById('numberInput');
    const userInput = escapeHtml(numberInput.value);
    console.log(userInput);

    numberInput.focus();

    // Making an error message.
    const errorMessage = newErrorMessageBin(answer, userInput);
    document.getElementById('errorArea').innerHTML = errorMessage;
    
    // Exits when the input was invalid.
    if (errorMessage) {
        return;
    }
    
    // Converting the input in order to use in the history message.
    const binaryDigit = 8;
    const zeroPaddedBin = userInput.padStart(binaryDigit, '0');
    const taggedBin = colorLeadingZero(zeroPaddedBin);
    console.log(zeroPaddedBin);
    console.log(taggedBin);
    
    const destinationRadix = 2;
    const userInputToDestRadix = parseInt(userInput, destinationRadix);
    console.log(userInputToDestRadix);
    
    const decimalDigit = 3;
    const spacePaddedDec = userInputToDestRadix.toString().padStart(decimalDigit, ' ').replace(' ', '&nbsp;');
    
    // Making a new history and updating the history with the new one.
    const sourceRadix = 10;
    const outputArea = document.getElementById('outputArea');
    const currentHistoryMessage = newHistory((userInputToDestRadix == answer), taggedBin, destinationRadix, spacePaddedDec, sourceRadix);
    const historyMessage = concatinateStrings(currentHistoryMessage, outputArea.innerHTML);
    console.log(currentHistoryMessage);
    console.log(historyMessage);
    outputArea.innerHTML = historyMessage;
    
    if (userInputToDestRadix == answer) {
        // Making the next question.
        let nextIndexNumber = 0;
        let nextAnswer = 0;

        console.log(last_answers);
        do {
            nextIndexNumber = getRandomBetween(0, 7);
            nextAnswer = Math.pow(2, nextIndexNumber);
            console.log(nextAnswer);
            console.log(last_answers.some((element) => element == nextAnswer));
        } while (last_answers.some((element) => element == nextAnswer));
        
        const nextHint = formatString(hint_format, [nextAnswer, nextIndexNumber]);
        console.log(nextHint);
        
        document.getElementById('questionSpan').innerText = nextAnswer;
        document.getElementById('hintArea').innerHTML = nextHint;
        numberInput.value = '';

        // Updating `lastAnswers`.
        // These numbers will not be used for the next question.
        const answersToKeep = 4;
        const lastAnswers = [nextAnswer].concat(last_answers).slice(0, answersToKeep);

        // Setting the next answer to the check button.
        document.getElementById('submitButton').onclick = function() { checkAnswerPot1(nextAnswer, hint_format, lastAnswers); return false; };
    }
}


function initPowerOfTwo1 () {
    // Initialization.
    const initIndexNumber = getRandomBetween(0, 7);
    const initAnswer = Math.pow(2, initIndexNumber);

    const hintFormat = '<details><summary>ヒント: </summary><p class="history-indented">{0}<sub>(10)</sub> = 2<sup>{1}</sup><br>10進法で2<sup>n</sup>になる数は、<br>2進法では1の後ろに0をn個つけます。</p></details>';
    const hint = formatString(hintFormat, [initAnswer, initIndexNumber]);

    const sourceRadix = 10;
    const destinationRadix = 2;

    document.getElementById('questionSpan').innerHTML = initAnswer;
    document.getElementById('srcRadix').innerHTML = '(' + sourceRadix + ')';
    document.getElementById('dstRadix').innerHTML = destinationRadix;
    document.getElementById('binaryRadix').innerHTML = '<sub>(' + destinationRadix + ')</sub>';
    document.getElementById('hintArea').innerHTML = hint;
    document.getElementById('submitButton').onclick = function() { checkAnswerPot1(initAnswer, hintFormat, [initAnswer]); return false;  };
}

File: src/endless-binary/PowerOfTwo1.fs
// taidalab
// https://github.com/taidalog/taidalab
// Copyright (c) 2022-2025 taidalog
// This software is licensed under the MIT License.
// https://github.com/taidalog/taidalab/blob/main/LICENSE
namespace Taidalab

open System
open Browser.Dom
open Taidalab.Number
open Taidalab.EndlessBinary
open Fermata.RadixConversion

module EndlessBinary =
    module PowerOfTwo1 =
        let help =
            """
            2<sup>n</sup> (0&le;n&le;7) の10進数から2進数への変換をエンドレスで練習できます。<br>
            2<sup>n</sup> の2進数を覚えると10進数からの変換を早く行えるので、まずはこのコースから始めてみてください。<br>
            ヒント付きなので、考え方も身に付けられます。
            """

        let hint answer =
            let indexNumber = Math.Log(double answer, 2.0) |> int

            $"""
            <details>
                <summary><h2>ヒント:</h2></summary>
                <p class="history-indented">
                    2<sup>n</sup> の数を2進法で表現するには、1 の後に 0 を n 個続けます。<br>
                    %d{answer}<sub>(10)</sub> は 2<sup>%d{indexNumber}</sup> なので、1 の後ろに 0 を %d{indexNumber} 個つけます。
                </p>
            </details>"""

        let question lastAnswers : int =
            newNumber (fun _ -> getRandomBetween 0 7 |> pown 2) (fun n -> List.contains n lastAnswers = false)

        let additional _ : unit = ()

        let exec' (lastNumbers: int list) (answer: int) : unit =
            Dec2Bin1.exec question hint newErrorMessageBin additional 4 lastNumbers answer

        let init () =
            document.title <- "2のn乗 - taidalab"

            let header = document.querySelector "header"
            header.innerHTML <- Content.Common.header
            header.className <- "power-of-two"

            (document.getElementById "hamburgerButton").onclick <-
                (fun _ ->
                    (document.querySelector "nav").classList.toggle "flagged" |> ignore
                    (document.getElementById "barrier").classList.toggle "flagged" |> ignore
                    (document.querySelector "main").classList.toggle "flagged" |> ignore)

            (document.getElementById "barrier").onclick <-
                (fun _ ->
                    (document.querySelector "nav").classList.remove "flagged" |> ignore
                    (document.getElementById "barrier").classList.remove "flagged" |> ignore
                    (document.querySelector "main").classList.remove "flagged" |> ignore)

            (document.querySelector "#headerTitle").innerHTML <-
                """<span>2のn乗 - </span><span translate="no">taidalab</span>"""

            (document.querySelector "main").innerHTML <- EndlessBinary.Course.main help "help-color power-of-two"
            (document.querySelector "#submitButton").className <- "power-of-two"
            (document.querySelector "#questionArea").innerHTML <- Content.Common.question

            Dec2Bin1.init' question hint additional 10 2 exec' EndlessBinary.keyboardshortcut
